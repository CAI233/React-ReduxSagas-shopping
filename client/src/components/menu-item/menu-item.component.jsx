import React from 'react';

import { withRouter } from 'react-router-dom';
import { MenuItemStyledContainer, Content, Title, Subtitle, BackgroundImage } from './menu-item.styles';

/**
 * 路由withRouter函数: 使当前页面下的组件, 可以直接访问"当前页面路由信息"( 完成笔记 )
 * 0. 引入: import { withRouter } from 'react-router-dom';
 * 1. 输出: export default withRouter( 组件名 );
 * 2. 使用: 路由属性名称 -> 直接传递给组件, 如下实战
 */
const MenuItem = ({ title,imageUrl,size,linkUrl,history,match }) => (

    <MenuItemStyledContainer 
        onClick={ ()=>history.push( `${match.url}${linkUrl}` ) } 
        className={ size ? `${size}` : `` }
    >
        <BackgroundImage 
            backgroundImage = {imageUrl}  
            className="background-image"
        >
        </BackgroundImage>
        <Content className="content">
            <Title>
                {title.toUpperCase()}
            </Title>
            <Subtitle>
                查看更多
            </Subtitle>
        </Content>
    </MenuItemStyledContainer>
);

export default withRouter(MenuItem);