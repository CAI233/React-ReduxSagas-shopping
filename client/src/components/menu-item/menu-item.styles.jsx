import styled,{css} from 'styled-components';
export const MenuItemStyledContainer = styled.div`
    min-width: 30%;
    height: 240px;
    flex: 1 1 auto;
    display: flex;
    align-items: center;
    justify-content: center;
    border: 1px solid black;
    margin: 0 7.5px 15px;
    overflow: hidden;

    // &.xxx相当于 .menu-item.xxx
    &.large{
        height:380px;
    }
  
    &:first-child {
      margin-right: 7.5px;
    }
  
    &:last-child {
      margin-left: 7.5px;
    }
    
    &:hover {
        cursor: pointer;

        // CSS-使标签缓缓变大( 完成笔记 )
        & .background-image {
            transform: scale(1.1);
            transition: transform 6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
        }
        
        & .content {
            opacity: 0.9;
        }
    }
`;
MenuItemStyledContainer.displayName = "MenuItemStyledContainer";
const backgroundImg = props => {
    if( props.backgroundImage ){
        return css`
            background-image: url(${ props.backgroundImage });
        `;
    }
}

export const BackgroundImage = styled.div`
    width:100%;
    height:100%;
    background: no-repeat;
    background-size:cover;
    background-position:center;
    ${ backgroundImg }
`;
BackgroundImage.displayName = "BackgroundImage";

export const Content = styled.div`
    height: 90px;
    padding: 0 25px;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    border: 1px solid black;
    position: absolute;
    background-color:#ddd ;
    opacity: 0.9;
`;
Content.displayName = "Content";

export const Title = styled.h1`
    font-weight: bold;
    margin-bottom: 6px;
    font-size: 22px;
    color: #4a4a4a;
`;
Title.displayName = "Title";

export const Subtitle = styled.span`
    font-weight: lighter;
    font-size: 16px;
`;
Subtitle.displayName = "Subtitle";