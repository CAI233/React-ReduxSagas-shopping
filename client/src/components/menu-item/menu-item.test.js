import React from 'react';
import { shallow } from 'enzyme';
import { MenuItem } from './menu-item.component';

describe(" 测试菜单列表显示 ",()=>{
    let wrapper,mockProps;
    beforeEach(()=>{
        mockProps = {
            title:"test",
            imageUrl: "test",
            size: "test",
            linkUrl: "test",
            history: {
                push: jest.fn()
            },
            match: {
                url: "test"
            }
        }
        wrapper = shallow( <MenuItem { ...mockProps } /> );
    });
    it("MenuItemSnapshot",()=>{
        expect(wrapper).toMatchSnapshot();
    });
    it("模拟单击跳转页面",()=>{
        wrapper.find("MenuItemStyledContainer").simulate('click');
        expect( mockProps.history.push ).toHaveBeenCalled();
    })

});