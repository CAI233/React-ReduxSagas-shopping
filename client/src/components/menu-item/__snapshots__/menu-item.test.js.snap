// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 测试菜单列表显示  MenuItemSnapshot 1`] = `ShallowWrapper {}`;
